apps/shared-app/src/client/product.tsx

```diff
@@ -15,10 +15,29 @@ export interface ProductProps {
 }
 
 export function Product({buy}: ProductProps): JSX.Element {
-  const [result, formAction] = ReactDOM.useFormState(buy, undefined);
+  const [formState, formAction] = ReactDOM.useFormState(buy, undefined);
+  const [result, setOptimisticResult] = React.useOptimistic(formState);
```

Comment:
IMO it's weird to me to pass the form state to useOptimistic. This isn't actually what you want to optimistically update, that is actually the product quantity.

Instead of returning the quantity from the mutation I think the recommended pattern is to fetch this in the server component and pass that to useOptimistic.

In other words: useOptimistic is for values, useFormState is for tracking the state of the action.

Reply:
Hm, you might have a point here. In a real-world example I would probably render a summary page based on the backend data (bought products) that I fetch in the server component, and not render the result of the buy action. But image this being and add-to-cart action instead of the actual buy action, and I just want to show a success or error toast. Wouldn't you base this on the result of the server action?

Reply:
That toast use case doesn't need to be both progressively enhanced and updated optimistically, so it's hard to say. Feels very different to what you're going for here.

Reply:
I definitely want to show a toast optimistically. And I would also appreciate it if the toast was shown even when the user clicked the button before the app was hydrated. Wouldn't you agree?

Reply:
I think that depends. In my opinion if you want to update something that you fetch from the server initially - passing formState to useOptimistic can be weird. But something like implementing form with lists of items in PE manner, then passing formState to useOptimistic can make it more responsive. Generally speaking, you can pass anything to useOptimistic 😄